http/relay/relay.go

```diff
@@ -168,11 +168,14 @@ retry:
 
 	// restart the server, try again
 	ctx.WithError(err).Error("network")
-	if err := p.Restart(); err != nil {
-		return nil, errors.Wrap(err, "restarting")
+
+	if restartErr := p.Restart(); restartErr != nil {
+		// We want to restart, but dont want to fail with that error.
+		// The real error was the network error that happened above
+		ctx.WithError(restartErr).Error("restarting")
 	}
 
-	goto retry
+	return nil, errors.Wrap(err, "network")
```

Comment:
What do you think about us checking the method instead? Then we can make more assumptions. Not sure what nginx and friends do there.

Reply:
My opinion on this stuff it to err on the side of correctness, at least until it can be configured or pluggable perhaps.

nginx had bad defaults for some time, but the defaults now are sane.  I'm not an expert, but it seems like you can configure them [here](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream).  They _only_ work for GET requests, but I think out of the box - they don't retry any application level errors.

I just checked out the docs for envoy, they seem to have configurable retries too [envoy retry docs](https://envoyproxy.github.io/envoy/configuration/http_filters/router_filter.html?highlight=retry#x-envoy-retry-on).

Allowing GETs to be retried X times might be ok - but IMO focusing on timeout behavior is much more important to get right / understand first.  Retries will exacerbate the timeout problems.

Maybe retry GET 5XX? maybe 409?

Reply:
Hmm yeah I'd think we would be safe to retry GET/HEAD/OPTIONS at least. That's true though, I overlooked that part, I'll open an issue to validating the backoff settings so it doesn't overflow

Reply:
Alright, I added to to retry those.